docs/source/how-to/selecting-the-collection-backend.rst

```diff
@@ -0,0 +1,131 @@
+
+- ``ones``
+- ``zeros``
+- ``empty``
+- ``full``
+- ``arange``
+
+The Dask-Array API can also dispatch the backend ``RandomState`` class to be used for random-number generation. This means all creation functions in ``dask.array.random`` are also dispatchable.
+
+The current set of dispatchable creation functions for Dask-Dataframe include:
+
+- ``from_dict``
+- ``read_parquet``
+- ``read_json``
+- ``read_orc``
+- ``read_csv``
+- ``read_hdf``
+
+As the backend-library disptaching system becomes more mature, this set of dispatchable creation functions is likely to grow.
+
+
+Defining a new collection backend
+---------------------------------
+
+**Warning**: Defining a custom backend is **not** yet recommended for most users and down-stream libraries. The backend-entrypoint system should still be treated as experimental.
+
+
+Dask curreontly exposes an `entrypoint <https://packaging.python.org/specifications/entry-points/>` under the group ``dask.backend`` to enable users and third-party libraries to develop and maintain backend implementations for Dask-Array and Dask-Dataframe. A custom Dask-Array backend should define a subclass of ``DaskArrayBackendEntrypoint`` (defined in ``dask.array.backends``), while a custom Dask-DataFrame backend should define a subclass of ``DaskDataFrameBackendEntrypoint`` (defined in ``dask.dataframe.backends``).
+
+For example, a cudf-based backend definition for Dask-Dataframe would look something like the `CudfBackendEntrypoint` definition below:
+
+
+.. code-block:: python
+   import dask.dataframe as dd
+   from dask.dataframe.backends import DaskDataFrameBackendEntrypoint
+   from dask.dataframe.dispatch import (
+      ...
+      make_meta_dispatch,
+      ...
+   )
+   ...
+
+   def make_meta_cudf(x, index=None):
+      return x.head(0)
+   ...
+
+   class CudfBackendEntrypoint(DaskDataFrameBackendEntrypoint):
+
+      def __init__(self):
+         # Register compute-based dispatch functions
+         # (e.g. make_meta_dispatch, sizeof_dispatch)
+         ...
+         make_meta_dispatch.register(
+            (cudf.Series, cudf.DataFrame),
+            func=make_meta_cudf,
+         )
+         # NOTE: Registration may also be outside __init__
+         # if it is in the same module as this class
+      ...
+
+      @staticmethod
+      def read_orc(*args, **kwargs):
+         from .io import read_orc
+
+         # Use dask_cudf version of read_orc
+         return read_orc(*args, **kwargs)
+      ...
+
+
+In order to expose this entrypoint as a ``dask.backend`` entrypoint, the necessary ``setup.cfg`` configuration in ``cudf`` (or ``dask_cudf``) would be as follows:
+
+.. code-block:: ini
+
+   [options.entry_points]
+   dask.backends =
+      cudf = <module-path>:CudfBackendEntrypoint
+
+
+Compute dispatch
+~~~~~~~~~~~~~~~~
```

Comment:
I really like this, thank you for writing it up (I had to mostly learn it by reading the code base).

There is a third way that things are dispatched, and it's arguably the most common in the code-base, though it's implicit. The NEP-18-based dispatching of `__array_function__` and its friends is, at least as far as I understand it, the most important way that different array backends get called, including cupy (correct me if this is wrong!).

I think if I were a new user, I'd look at the "compute-based dispatch functions" that you've listed below and think "That's a weird list, those don't seem like the most important functions at all!". But the reason those dispatch functions are there is that dask is mostly doing some additional custom logic which is not well-served by the numpy-based dispatching (again, correct me if you have a different perspective here!).

So again, I really like this new section, but I also think we'll probably want to expand on it more (though it's likely outside of the scope of this PR).

Reply:
This is a good point! I added a note about the importance of `__array_function__` to the beginning of this section. Maybe that is sufficient for now?

Reply:
I think that note is great, thanks for adding it.